File: figures/09-string-matching/snippet-rabin-karp.java
    public static List<Integer> rabinKarp(String text, String pattern, long base, long modulus) {
        int n = text.length(), m = pattern.length();
        if (m == 0) return List.of(0);
        if (m > n) return List.of();
        long h = 1, pHash = 0, tHash = 0;
        for (int j = 1; j < m; j++) h = mulMod(h, base, modulus);
        for (int j = 0; j < m; j++) {
            pHash = addMod(mulMod(pHash, base, modulus), code(pattern.charAt(j)), modulus);
            tHash = addMod(mulMod(tHash, base, modulus), code(text.charAt(j)), modulus);
        }
        List<Integer> out = new ArrayList<>();
        for (int s = 0; s + m <= n; s++) {
            if (pHash == tHash && equalsAt(text, pattern, s)) out.add(s);
            if (s + m < n) {
                long leading = mulMod(code(text.charAt(s)), h, modulus);
                long remainder = subMod(tHash, leading, modulus);
                tHash = addMod(mulMod(base, remainder, modulus), code(text.charAt(s + m)), modulus);
            }
        }
        return out;
    }

File: figures/09-string-matching/snippet-rabin-karp.py
def rabin_karp(text, pattern, base, modulus):
    n, m = len(text), len(pattern)
    if m == 0:
        return [0]
    if m > n:
        return []
    h = 1
    for _ in range(1, m):
        h = _mul_mod(h, base, modulus)
    p_hash = t_hash = 0
    for j in range(m):
        p_hash = _add_mod(_mul_mod(p_hash, base, modulus), _code(pattern[j]), modulus)
        t_hash = _add_mod(_mul_mod(t_hash, base, modulus), _code(text[j]), modulus)
    result = []
    for s in range(n - m + 1):
        if p_hash == t_hash and text[s:s + m] == pattern:
            result.append(s)
        if s < n - m:
            leading = _mul_mod(_code(text[s]), h, modulus)
            t_hash = _add_mod(_mul_mod(base, _sub_mod(t_hash, leading, modulus), modulus),
                               _code(text[s + m]), modulus)
    return result
